ports/qemu-arm/Makefile

```diff
@@ -90,65 +59,107 @@ endif
+# bare metal RISC-V toolchain with Picolibc rather than Newlib, and the default
+# is "nosys" so a value must be provided.  To avoid having per-distro
+# workarounds, always select Picolibc if available.
+PICOLIBC_SPECS = $(shell $(CC) --print-file-name=picolibc.specs)
+ifeq ($(PICOLIBC_SPECS),picolibc.specs)
+# Picolibc was not found.
+SPECS_FRAGMENT =
+else
+SPECS_FRAGMENT = --specs=$(PICOLIBC_SPECS)
+CFLAGS += $(SPECS_FRAGMENT)
+endif
+
+LD = $(CC)
+
+AFLAGS += -mabi=$(RV32_ABI) -march=$(RV32_ARCH)
+CFLAGS += $(AFLAGS)
+LDFLAGS += -mabi=$(RV32_ABI) -march=$(RV32_ARCH) -T $(LDSCRIPT) -Wl,-EL
+LDFLAGS += $(SPECS_FRAGMENT) -Wl,--gc-sections -Wl,-Map=$(@:.elf=.map)
+
+SRC_C += \
+	mcu/rv32/interrupts.c \
+	mcu/rv32/startup.c \
+
+SRC_BOARD_O += mcu/rv32/entrypoint.o
+
+endif
+
+SRC_C += \
+	main.c \
 	uart.c \
 	mphalport.c \
 	shared/libc/string0.c \
 	shared/readline/readline.c \
 	shared/runtime/interrupt_char.c \
 	shared/runtime/pyexec.c \
-	shared/runtime/semihosting_arm.c \
 	shared/runtime/stdout_helpers.c \
 	shared/runtime/sys_stdio_mphal.c \
 
-SRC_RUN_C = \
-	main.c \
-
 LIB_SRC_C += $(SRC_LIB_LIBM_C)
 LIB_SRC_C += $(SRC_LIB_LIBM_SQRT_SW_C)
 
-OBJ_COMMON =
-OBJ_COMMON += $(PY_O)
-OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_COMMON_C:.c=.o))
-OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_BOARD_O))
-OBJ_COMMON += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
-
-OBJ_RUN =
-OBJ_RUN += $(addprefix $(BUILD)/, $(SRC_RUN_C:.c=.o))
-
-ALL_OBJ_RUN = $(OBJ_COMMON) $(OBJ_RUN)
-
-OBJ_TEST =
-OBJ_TEST += $(addprefix $(BUILD)/, $(SRC_TEST_C:.c=.o))
-
-ALL_OBJ_TEST = $(OBJ_COMMON) $(OBJ_TEST)
-
-# All object files, needed to get dependencies correct
-OBJ = $(OBJ_COMMON) $(OBJ_RUN) $(OBJ_TEST)
+OBJ += $(PY_O)
+OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
+OBJ += $(addprefix $(BUILD)/, $(SRC_BOARD_O))
+OBJ += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
 
 # List of sources for qstr extraction
-SRC_QSTR += $(SRC_COMMON_C) $(SRC_RUN_C) $(LIB_SRC_C)
+SRC_QSTR += $(SRC_C) $(LIB_SRC_C)
 
 all: $(BUILD)/firmware.elf
 
 .PHONY: repl
 repl: $(BUILD)/firmware.elf
 	$(ECHO) "Use machine.reset() to exit"
-	qemu-system-arm $(QEMU_ARGS) -serial mon:stdio -kernel $<
+	$(QEMU_SYSTEM) $(QEMU_ARGS) -serial mon:stdio -kernel $<
 
 .PHONY: run
 run: $(BUILD)/firmware.elf
-	qemu-system-arm $(QEMU_ARGS) -serial pty -kernel $<
+	$(QEMU_SYSTEM) $(QEMU_ARGS) -serial pty -kernel $<
+
+# `make debug` will block QEMU until a debugger is connected to port 1234.
+.PHONY: debug
+debug: $(BUILD)/firmware.elf
+	$(QEMU_SYSTEM) $(QEMU_ARGS) -serial mon:stdio -S -s -kernel $<
 
 .PHONY: test
 test: $(BUILD)/firmware.elf
 	$(eval DIRNAME=ports/$(notdir $(CURDIR)))
-	cd $(TOP)/tests && ./run-tests.py --target qemu-arm --device execpty:"qemu-system-arm $(QEMU_ARGS) -serial pty -kernel ../$(DIRNAME)/$<" $(TESTS_EXCLUDE)
+	cd $(TOP)/tests && ./run-tests.py --target qemu-arm --device execpty:"$(QEMU_SYSTEM) $(QEMU_ARGS) -serial pty -kernel ../$(DIRNAME)/$<" $(TESTS_EXCLUDE)
```

Comment:
I'd add another argument to this for additional user arguments for `run-tests.py`, or rename `TESTS_EXCLUDE` into something else and have it kick in only if it's not defined.

Reply:
OK, how about adding `$(RUN_TESTS_EXTRA)`?  If it's not defined it won't do anything.  The user can set it from the command line, eg to `-d basics`.

Reply:
That would work, too.  Would it help if these extra arguments (along with the other targets) are documented or are niche enough that if you need them you'd have probably figured them out already?

Reply:
I can document the extra args, and the possible options for `BOARD=`.

Reply:
I added `RUN_TEST_EXTRA`, and documented it in the README, along with a list of available boards.